packages/react-dom/src/server/ReactPartialRenderer.js

```diff
@@ -87,38 +86,40 @@ if (__DEV__) {
     return describeComponentFrame(name, source, ownerName);
   };
 
-  currentDebugStack = null;
-  currentDebugElementStack = null;
-  setCurrentDebugStack = function(stack: Array<Frame>) {
-    const frame: Frame = stack[stack.length - 1];
-    currentDebugElementStack = ((frame: any): FrameDev).debugElementStack;
+  // ReactDOMServer is reentrant, so there may be more than one top frame.
+  // This is why we keep an array of them in `currentDebugFrames`.
+  pushCurrentDebugFrame = function(frame: FrameDev) {
     // We are about to enter a new composite stack, reset the array.
-    currentDebugElementStack.length = 0;
-    currentDebugStack = stack;
-    ReactDebugCurrentFrame.getCurrentStack = getStackAddendum;
-  };
-  pushElementToDebugStack = function(element: ReactElement) {
-    if (currentDebugElementStack !== null) {
-      currentDebugElementStack.push(element);
+    frame.debugElementStack.length = 0;
+    currentDebugFrames.push(frame);
+    if (currentDebugFrames.length === 1) {
+      ReactDebugCurrentFrame.getCurrentStack = getStackAddendum;
     }
   };
-  resetCurrentDebugStack = function() {
-    currentDebugElementStack = null;
-    currentDebugStack = null;
-    ReactDebugCurrentFrame.getCurrentStack = null;
+  pushElementToDebugFrame = function(element: ReactElement) {
+    // A top frame should always exist.
+    const frame = currentDebugFrames[currentDebugFrames.length - 1];
+    frame.debugElementStack.push(element);
+  };
+  popCurrentDebugFrame = function() {
+    currentDebugFrames.pop();
+    if (currentDebugFrames.length === 0) {
+      ReactDebugCurrentFrame.getCurrentStack = null;
+    }
   };
   getStackAddendum = function(): null | string {
-    if (currentDebugStack === null) {
+    const currentDebugFrame = currentDebugFrames[currentDebugFrames.length - 1];
+    if (currentDebugFrame === null) {
```

Comment:
Just curious, but when would this happen? I don't see anywhere we push `null` to this stack.

Reply:
Oops. Shouldn't happen, but I'll change it to test for array length instead. It's for exceptional case where we accidentally call `getStackAddendum` too late.